### ISSUE TITLE
Fix violations in local IntelliJ 2017.3

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#5409]
My local IntelliJ on windows shows 50+ errors and 250+ warnings when running the inspection file on checkstyle's code.

![intellij](https://user-images.githubusercontent.com/5427943/34459211-64325314-edb7-11e7-97ec-af2eb8149a50.png)

We should fix these where possible. We can't suppress them since TeamCity will complain that they aren't needed since it isn't producing a violation on them now.

**Comment by romani** [comment]
We can suppress all new violations, and quickly switch to new version and then step by step resolve saved list of suppressions.

Does it make sense ?

**Comment by romani** [comment]
Please share report somehow , you have too much violations, I need to recheck that you correctly setup inspections

**Comment by rnveach** [comment]
Here is my report: http://rveach.no-ip.org/checkstyle/regression/reports/issue5409.html

I followed directions on our website to setup local inspection.
Even if it is slightly wrong, I am only fixing the ones I agree with. The first PR had a lot of valid fixes, especially the regular expression ones.

**Comment by romani** [comment]
issue is approved, 

Is NOT ok to fix: Declaration redundancy . But such violations looks familiar from summer, we somehow managed to suppress them or escape by mean of configuration.

**Comment by rnveach** [comment]
> Is NOT ok to fix: Declaration redundancy . 

Why can't some of `Declaration access can be weaker inspection` be fixed for non-api code?

For example, `class XMLLogger WARNING Can be protected`,
inner class is `private`
https://github.com/checkstyle/checkstyle/blob/90f20e09869c71eb22190ad9c1d46d5deec324a5/src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java#L345
so it doesn't make sense for methods to be `public`
https://github.com/checkstyle/checkstyle/blob/90f20e09869c71eb22190ad9c1d46d5deec324a5/src/main/java/com/puppycrawl/tools/checkstyle/XMLLogger.java#L357

**Comment by romani** [comment]
In my experience "protected" always lead to hacks and problems in design, as it presume that there are friends of class who can access internals. Class should live a independent item irrespective in what package it is declared - https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html .
A bit less evil than "no modifier". Yes I do know that there exceptions and some time it might be convenient, but this convenient might be just another laziness to think about better design.

> class XMLLogger WARNING Can be protected

No, if smb need to reuse it, it should become top level public class. If nobody need it, it should be hidden.

> so it doesn't make sense for methods to be public

to my mind it does. As class is ready to become top level class at any time.
We can make it "private" top class will have access to them no matter what, but class will looks strange.

**Comment by rnveach** [comment]
> it should become top level public class. 

The violation was just talking about the methods inside the class. It wasn't asking to change the modifier for the class. The class is already private, so changing protected wouldn't make it weaker, but stronger.

> to my mind it does

ok.

**Comment by rnveach** [comment]
Here is the final results from my original report:

`Unused 'catch' parameter inspection   1 warning`
Done.
`Overridable method called during object construction inspection   8 errors`
Done.
`Unsafe lazy initialization of 'static' field inspection   2 errors`
Done.
`Zero‑length array allocation inspection   1 warning`
Done.
`RegExp group   145 warnings`
Done.

`Declaration access can be weaker inspection   106 warnings`
These can't be done as discussed.
`Redundant throws clause inspection   25 errors`
The rest of these can't be done.
`Redundant suppression inspection   10 errors`
These can't be done until TC says they aren't needed.
`Abstract class which has only one direct inheritor inspection   2 errors`
Won't be done as we should keep all classes together.
`Class naming convention inspection   6 errors`
Won't be done as these are inner classes and shouldn't be done.
`Reflective access to nonexistent/not visible class member inspection   1 warning`
This looks like a false positive as tests would fail if we couldn't find the member.

**Comment by romani** [comment]
in TC there is only 2017.2.4 , so ideally we should satisfy this version of inspections to make CI pass:
[image](https://user-images.githubusercontent.com/812984/34640133-f50eeaac-f2a1-11e7-853e-8f84cd38fa2a.png)

**Comment by romani** [comment]
violations from 2017.2.6:
![image](https://user-images.githubusercontent.com/812984/34640342-a809c7c8-f2a5-11e7-8e19-a88b2b15201f.png)

**Comment by rnveach** [comment]
> We need to change issue title to reference 2017.2.4(6) 

@romani since you are discussing updating TC, lets move new version to existing issue https://github.com/checkstyle/checkstyle/issues/5069 and leave this one as what is on my local.
Once the last 2 PRs are finished, this issue can be closed anyways until we update TC.

> violations from 2017.2.6:

I thought i fixed all RegExp ones. Some of the others are similar to what I got.

**Comment by romani** [comment]
Looks like I lost a track of issues, that is good that we have such issue already

**Comment by romani** [comment]
Are we ready to upgrade TC, please recheck what is left to be fixed

**Comment by rnveach** [comment]
@romani Nothing is left to be fixed from original report.
Lets close this issue and do upgrade in issue mentioned above (https://github.com/checkstyle/checkstyle/issues/5069).

**Comment by romani** [comment]
We can not update to 2017.2.4(6), it still shows violations.

**Comment by rnveach** [comment]
@romani There is still nothing we can do in this issue.
Violations remaining are specified at https://github.com/checkstyle/checkstyle/issues/5409#issuecomment-355715454 . They can't be fixed for reasons I gave (unless you want to overrule my reasoning), and they can't be suppressed because old TC would complain the suppression isn't needed.
Violations from upgrading TC must be fixed in #5069 with the upgrade.

**Comment by romani** [comment]
@rnveach , I updated TC to 2017.2.6 , most recent version in TeamCity for now.

If you see no violation on your side - it is better to close this issue.

**Comment by romani** [comment]
after my PR, there should be left only removal of suppressions, I used Idea 2017.3.4 .
![image](https://user-images.githubusercontent.com/812984/36637282-bc345f42-198d-11e8-94f4-5e269f141beb.png)

but we will wait till this version appear in TeamCity CI.

**Comment by romani** [comment]
I am closing issue , as nothing is left for now (2017.3.4 version), except for redundant suppressions.
They will be easily resolved ones 2017.3.4 appear in TeamCity CI.

**Comment by romani** [comment]
just to keep notes of what I did today ...

I added skip for build trigger if changes is done in others CIs files:
```
-:*.yml
-:.ci/**
-:.github/**
-:codeship-*
-:circleci/*
-:LICENSE*
-:README*
-:release.sh
-:RIGHTS.*
```

![image](https://user-images.githubusercontent.com/812984/40404677-f0f7f274-5e0d-11e8-8293-3526c2823ed4.png)
